### PR TITLE
Add Codecov

### PR DESCRIPTION
**Pull request by kvz**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -66,6 +66,37 @@ jobs:
           node-version: ${{ matrix.node }}
       - run: corepack yarn
       - run: corepack yarn vitest run --coverage ./test/unit
+      - name: Upload coverage reports
+        if: matrix.node == 22 # Only upload coverage from the latest Node.js version
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-reports
+          path: coverage/
+      - name: Upload coverage to Codecov
```
Comment (remcohaszing) [inline .github/workflows/ci.yml:75]: Personally I’m not a fan of using the `name` property too much. It acts as a label, which defaults to the action name. IMO `codecov/codecov-action@v4` is more descriptive than a custom label.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -66,6 +66,37 @@ jobs:
...
+  coverage:
+    needs: vitest
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/download-artifact@v4
+        with:
+          name: coverage-reports
+          path: coverage
+      - name: Check coverage thresholds
```
Comment (remcohaszing) [inline .github/workflows/ci.yml:94]: In case of a multi-line shell script I do think the name is useful.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -66,6 +66,37 @@ jobs:
...
+        run: |
+          if [ ! -f coverage/coverage-summary.json ]; then
+            echo "No coverage report found"
+            exit 1
+          fi
```
Comment (remcohaszing) [inline .github/workflows/ci.yml:99]: This doesn’t actually check the thresholds. Also doesn’t Vitest already fail if the threshold isn’t met?

File: vitest.config.mjs [hunk 4]
```diff
@@ -3,9 +3,22 @@ import { defineConfig } from 'vitest/config'
...
+        branches: 2,
+        functions: 0,
+        lines: 2,
+        perFile: true,
+      },
+      clean: true,
+      reportsDirectory: './coverage',
     },
     globals: true,
+    environment: 'node',
```
Comment (remcohaszing) [inline vitest.config.mjs:22]: I think these options are the default. So why specify them?

```suggestion
    },
    globals: true,
```

**Review by remcohaszing — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (mifi) [reply .github/workflows/ci.yml:75]: I agree that name is not needed when the `uses` is descriptive, however in this case I don't think `codecov/codecov-action@v4` is more descriptive than `Upload coverage to Codecov`. it's not obvious that `codecov/codecov-action@v4` does just an upload.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -66,6 +66,37 @@ jobs:
           node-version: ${{ matrix.node }}
       - run: corepack yarn
       - run: corepack yarn vitest run --coverage ./test/unit
+      - name: Upload coverage reports
```
Comment (mifi) [inline .github/workflows/ci.yml:69]:
```suggestion
      - name: Upload coverage reports artifact
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -66,6 +66,37 @@ jobs:
...
+        if: matrix.node == 22 # Only upload coverage from the latest Node.js version
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-reports
+          path: coverage/
+      - name: Upload coverage to Codecov
+        if: matrix.node == 22 # Only upload coverage from the latest Node.js version
```
Comment (mifi) [inline .github/workflows/ci.yml:76]: maybe we can move this into the `coverage` job, then we don't have to check for node version again also

**Pull request merged**